Source/CarthageKit/DiagnosticResolver.swift

```diff
@@ -0,0 +1,197 @@
+}
+
+/**
+ Resolver which logs all dependencies it encounters, and optionally stores them in the specified local repository.
+ */
+public final class DiagnosticResolver: ResolverProtocol {
+    private let localRepository: LocalRepository
+    private let versionsForDependency: (Dependency) -> SignalProducer<PinnedVersion, CarthageError>
+    private let resolvedGitReference: (Dependency, String) -> SignalProducer<PinnedVersion, CarthageError>
+    private let dependenciesForDependency: (Dependency, PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError>
+
+    private var versionSets = [Dependency: [PinnedVersion]]()
+    private var handledDependencies = Set<PinnedDependency>()
+
+    public var ignoreErrors: Bool = false
+
+    // Specify mappings to anonimize private dependencies (which may not be disclosed as part of the diagnostics)
+    public var dependencyMappings: [Dependency: Dependency]?
+
+    public let diagnosticResolverEvents: Signal<DiagnosticResolverEvent, NoError>
+    private let diagnosticResolverEventPublisher: Signal<DiagnosticResolverEvent, NoError>.Observer
+
+    private enum DiagnosticResolverError: Error {
+        case versionRetrievalFailure(message: String)
+        case dependencyRetrievalFailure(message: String)
+    }
+
+    public convenience init(versionsForDependency: @escaping (Dependency) -> SignalProducer<PinnedVersion, CarthageError>, dependenciesForDependency: @escaping (Dependency, PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError>, resolvedGitReference: @escaping (Dependency, String) -> SignalProducer<PinnedVersion, CarthageError>) {
+        self.init(versionsForDependency: versionsForDependency,
+                  dependenciesForDependency: dependenciesForDependency,
+                  resolvedGitReference: resolvedGitReference,
+                  localRepository: LocalRepository(directoryURL: URL(fileURLWithPath: "/tmp")))
+    }
+
+    public init(
+        versionsForDependency: @escaping (Dependency) -> SignalProducer<PinnedVersion, CarthageError>,
+        dependenciesForDependency: @escaping (Dependency, PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError>,
+        resolvedGitReference: @escaping (Dependency, String) -> SignalProducer<PinnedVersion, CarthageError>,
+        localRepository: LocalRepository) {
+        self.versionsForDependency = versionsForDependency
+        self.dependenciesForDependency = dependenciesForDependency
+        self.resolvedGitReference = resolvedGitReference
+        self.localRepository = localRepository
+
+        let (signal, observer) = Signal<DiagnosticResolverEvent, NoError>.pipe()
+        diagnosticResolverEvents = signal
+        diagnosticResolverEventPublisher = observer
+    }
+
+    public func resolve(
+        dependencies: [Dependency: VersionSpecifier],
+        lastResolved: [Dependency: PinnedVersion]? = nil,
+        dependenciesToUpdate: [String]? = nil
```

Comment:
this argument is not used. What do you think or asserting if not nill?

Reply:
Removed lastResolved, dependenciesToUpdate. The class is now not an implementation of Resolver anymore.